handler/openid/flow_device_token.go

```diff
@@ -0,0 +1,64 @@
+// Copyright © 2024 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package openid
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/fosite"
+	"github.com/ory/x/errorsx"
+)
+
+func (c *OpenIDConnectDeviceHandler) HandleTokenEndpointRequest(ctx context.Context, requester fosite.AccessRequester) error {
+	return errorsx.WithStack(fosite.ErrUnknownRequest)
+}
+
+func (c *OpenIDConnectDeviceHandler) PopulateTokenEndpointResponse(ctx context.Context, requester fosite.AccessRequester, responder fosite.AccessResponder) error {
+	if !c.CanHandleTokenEndpointRequest(ctx, requester) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	if !requester.GetClient().GetGrantTypes().Has(string(fosite.GrantTypeDeviceCode)) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHint("The OAuth 2.0 Client is not allowed to use the authorization grant \"urn:ietf:params:oauth:grant-type:device_code\"."))
+	}
+
+	deviceCode := requester.GetRequestForm().Get("device_code")
+	signature, err := c.DeviceCodeStrategy.DeviceCodeSignature(ctx, deviceCode)
+	ar, err := c.OpenIDConnectRequestStorage.GetOpenIDConnectSession(ctx, signature, requester)
+	if errors.Is(err, ErrNoSessionFound) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest.WithWrap(err).WithDebug(err.Error()))
+	}
+	if err != nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	if !ar.GetGrantedScopes().Has("openid") {
+		return errorsx.WithStack(fosite.ErrMisconfiguration.WithDebug("An OpenID Connect session was found but the openid scope is missing, probably due to a broken code configuration."))
+	}
+
+	session, ok := ar.GetSession().(Session)
+	if !ok {
+		return errorsx.WithStack(fosite.ErrServerError.WithDebug("Failed to generate id token because session must be of type fosite/handler/openid.Session."))
+	}
+
+	claims := session.IDTokenClaims()
+	if claims.Subject == "" {
+		return errorsx.WithStack(fosite.ErrServerError.WithDebug("Failed to generate id token because subject is an empty string."))
+	}
+
+	claims.AccessTokenHash = c.GetAccessTokenHash(ctx, requester, responder)
+
+	idTokenLifespan := fosite.GetEffectiveLifespan(requester.GetClient(), fosite.GrantTypeDeviceCode, fosite.IDToken, c.Config.GetIDTokenLifespan(ctx))
+	return c.IssueExplicitIDToken(ctx, idTokenLifespan, ar, responder)
+}
+
```

Comment:
In flow_explicit_token, we are deleting the openid connect session on use:


	err = c.OpenIDConnectRequestStorage.DeleteOpenIDConnectSession(ctx, authorizeCode)
	if err != nil {
		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
	}

I'm not sure if we need to do that here as well, or if this is covered by flow_explicit_token. However, as I understand it, flow_explicit_token is not called when performing an device auth code grant? So is this call missing?

Reply:
@nsklikas could you please check this comment?

Reply:
You were right that it was missing, we either missed it or something had changed after we started working on this feature. Your suggestion has been added (see https://github.com/ory/fosite/pull/826/files#diff-408286bd02ab957e8c252a7ae6beb1c848873cb66ffba56ca0cf29dbe0658bd6R52)